### PR TITLE
Fix process leaks in tool tests

### PR DESCRIPTION
**Pull request by dnfield**
The integration tests in the tool invoke Flutter to:

- Create project(s) in temp directories
- Run those projects directly via flutter_tester
- Make sure that you can attach a debugger, evaluate expressions, etc.

These processes leak.  The shell script/batch script files are not set up to wait for the `dart` process they invoke to finish, and thus they can exit (and complete the `process.exitCode` future) before the Dart VM process they fork actually finishes. On Windows, this is particularly bad, since the leaked process(es) end up holding an exclusive lock on the temp directory, meaning we fail to clean it up. We've been swallowing that under the assumption that it was a permissions issue, but it appears to not be a permissions issue and just be an issue of leaked processes.

This patch adds some additional waits/attempts to terminate the Dart VM(s) that the tests spawn. In cases where we cannot be sure that the spawned VM has terminated, it adds some additional OS specific waiting to check that the PID the VM created is no longer active (on Windows in `tasklist` and on POSIX in response to `kill -0`).

We should be able to further improve the test_driver logic if we can get https://github.com/dart-lang/sdk/issues/40759 resolved.  Alternatively, we could look at making `flutter.bat` wait for Dart to exit when it launches, but I'm a bit scared to touch that logic since it has some comments about how it has to work in a bunch of scenarios I'm not sure we have tests for (/cc @goderbauer).  The Dart feature would be useful regardless of this particular instance of usage though.

### CODE REVIEW

File: packages/flutter_tools/test/src/common.dart [hunk 1]
```diff
@@ -25,14 +25,7 @@ export 'package:test_core/test_core.dart' hide TypeMatcher, isInstanceOf; // Def
...
 void tryToDelete(Directory directory) {
-  // This should not be necessary, but it turns out that
-  // on Windows it's common for deletions to fail due to
-  // bogus (we think) "access denied" errors.
-  try {
-    directory.deleteSync(recursive: true);
-  } on FileSystemException catch (error) {
-    print('Failed to delete ${directory.path}: $error');
-  }
+  directory.deleteSync(recursive: true);
```
Comment (dnfield) [inline packages/flutter_tools/test/src/common.dart]: I'd like to leave this for now in case we see flakes around this pop up again.  Should I add a TODO to remove it?

Comment (dnfield) [reply packages/flutter_tools/test/src/common.dart]: The _actual_ reason these files were failing to delete was because we had processes that weren't being killed before trying to delete files in use, which on Windows causes exceptions.

File: packages/flutter_tools/test/integration.shard/test_driver.dart [hunk 2]
```diff
@@ -173,15 +174,55 @@ abstract class FlutterTestDriver {
...
+    print('killing forcefully');
     _debugPrint('Sending SIGKILL to $_processPid..');
     ProcessSignal.SIGKILL.send(_processPid);
+    await _ensureDead(_processPid, 10);
     return _process.exitCode;
   }
 
+  // TODO(dnfield): This is racy. If dart-lang/sdk#40759 can be resolved, we
+  // should remove this.
+  Future<void> _ensureDead(int pid, int tries) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/integration.shard/test_driver.dart]: uber nit: shouldn't be called `ensureX` if it can't ensure it.

Comment (dnfield) [reply packages/flutter_tools/test/integration.shard/test_driver.dart]: Renamed to `tryWaitForPidDeath`, and moved it to test_utils.dart so it  can be reused in the daemon_mode_test for the same reason.

Comment (jonahwilliams) [reply packages/flutter_tools/test/integration.shard/test_driver.dart]: SGTM

File: packages/flutter_tools/test/src/common.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/test/src/common.dart]: Could you link to the Dart SDK bug?

**Comment by jonahwilliams** [comment]
This seems reasonable to be, but I'd wait for @zanderso to take a look too

File: packages/flutter_tools/test/src/common.dart (hunk 1 above)
Comment (dnfield) [reply packages/flutter_tools/test/src/common.dart]: I'm not sure if that's the right bug to link here - if anything I'd say file a new bug to track removing this code once we're convinced the current change hasn't re-introduced old flakes.

**Comment by zanderso** [comment]
Initial thoughts:
1. Please expand the PR description to explain what problem this is solving (or link to the github issue), and why it solves it, and if there are any remaining issues.
2. I think we need to take a step back and think about what is being bought by going through the shell script indirection here rather than invoking Dart directly. It seems like if the test driver had a `Process` object for the VM, then some difficulty here could be avoided, and we wouldn't need a feature request against `killPid`.

File: dev/ci/docker_linux/Dockerfile [hunk 3]
```diff
@@ -27,7 +27,8 @@ RUN apt-get install -y --no-install-recommends \
   unzip \
   apt-transport-https \
   ca-certificates \
-  gnupg
+  gnupg \
+  procps
```
Comment (dnfield) [inline dev/ci/docker_linux/Dockerfile]: @christopherfujino  - should I split this into a separate PR?

This is needed so that `kill` is available in the container.

**Comment by dnfield** [comment]
@zanderso - I updated the PR description.

The closest bug I can think of to this is https://github.com/flutter/flutter/issues/29141. The resolution of that issue was to make an outer harness just kill processes at the end of it.  We can't quite do that in these tests.

As to whether it would be good enough to avoid the `flutter.bat`/`flutter` shell scripts, I'd leave that to @Hixie - I believe he wrote most of these tests.

**Comment by dnfield** [comment]
We also might still have a real bug such that in certain debugging scenarios, the Windows VM leaks. That seems to be the case in these tests, without explicitly killing it.

**Comment by dnfield** [comment]
Hm.  This seemed to work locally on a Windows box, but I think the Windows instance we're using is just slower...  Not sure why the process is still sticking around.

**Comment by dnfield** [comment]
Discussed offline.  This approach will still be flaky, and it's not clear that it's any less flaky than the current approach (even though it may avoid flakes in many common cases).

I'm going to file an issue specifically about some of the things I've discovered here.  I'll leave this open for now - hopefully @hixie might have some good input on this as the author of many of these tests.

**Comment by dnfield** [comment]
I'm going to close this - it's not ready as is and I don't have a good sense of when it will be.
